README
Tastydir - an AJAX file manager and directory listing

INSTALLATION
	Simply upload everything to a directory of your choice.
	
DOCUMENTATION
	Look at the /doc directory in your browser for detailed documentation.
	
PERMISSION PROBLEMS?
	Your PHP scripts may be running as the 'nobody' user (you can check this in the info tooltip in Tastydir).
	This is common on many shared hosts, and it means that Tastydir won't be able to edit files created with FTP.
	SOLUTION: when you upload files with FTP or via cPanel, make sure you chmod them to 777.

CONTACT
	Found a bug? Have a suggestion? Contact me.
	Contact:	vlad@vladh.net

CREDITS
	Copyright Vlad Harbuz (vladh.net)

	Additional credits:
		Icons:
			Silk icons 					cc-by 2.5		http://www.famfamfam.com/lab/icons/silk/
			LED icons  					-				http://led24.de/iconset/
		jQuery:
			jQuery						MIT license		http://jquery.com/
			jQuery hashchange event		MIT, GPL		http://benalman.com/projects/jquery-hashchange-plugin/
			Tipsy						MIT				http://plugins.jquery.com/project/tipsy
		Other:
			jQuery UI CSS framework		MIT, GPL		http://jqueryui.com/themeroller/

CHANGELOG
=== Version 1.2 ===
1200:
	Submitted to Envato (CodeCanyon).
1205:
	Added changelog (lol).
	Minor CSS update on buttons.
	Prevented direct access to common.php and install.php.
1210:
	Submission to Envato soft-rejected, adding more documentation.
	Added documentation in /doc.
	Small optimizations.
1212:
	Approved by Envato.
	Fixed bug where trying to upload a file without actually selecting one hid the form.
	Fixed a bug where the file override warning in the upload function would show an incorrect or blank name.
1213:
	Changed vladh.net/projects/tastydir link to tastydev.net/#tastydir.
	Revision number now visible by hovering over version number.
	Changed bottom border on tooltip links to match abbr and be more subtle.
1214:
	Fixed typo ("755 or 755" instead of "755 or 777") in install.php.
	Fixed a bug where files with quotes/apostrophes in their names would cause problems.
		When working with these files, the user would be informed that the file doesn't exist.
1215:
	Added support for changing the home directory (top of common.php).
1216:
	PHP4 and early PHP5 now work properly ( they don't have json_encode :( ).
	Tastydir no longer incorrectly trips some antivirus programs.
	Fixed a silly "path not defined" problem in common.php.
1217:
	Fixed a problem where POSIX functions not being available on some Linux systems
		caused index.php to show up as blank due to a fatal error.
	Added error_reporting(E_ALL), to make errors easier to identify.
1250:
	Added two heavily requested features:
	- unless you're logged in, the top address bar doesn't show the full path, but the relative path.
	- added previous/next buttons on images for an image gallery-style effect.
1251:
	Password input in install.php is now an actual password input.
	When editing files, saving the file no longer closes the dialog, making it more suitable for programmers.
1252:
	Made all MIME types work properly.

~vladh (vladh.net)

_tastydir/do.php
<?php

// Tastydir is copyright Vlad Harbuz (vladh.net)

define('TASTYDIR_VERSION','1.2');
define('TASTYDIR_SUBVERSION','1251');

error_reporting(0);

require_once 'common.php';
if(!file_exists('settings.php')){
	echo 'settings.php doesn\'t exist! Please run the install script.';
	die();
}else{
	require_once 'settings.php';
}

if(empty($homedir)){
	$path=dirname(dirname($_SERVER['SCRIPT_FILENAME']));
}else{
	$path=$homedir;
}

// check if logged in
$auth['loggedin']=false;
if($auth['use']){
	if(!empty($_COOKIE['tastydir_auth']) && $_COOKIE['tastydir_auth']==sha1($auth['pass'])){
		$auth['loggedin']=true;
	}
}else{
	$auth['loggedin']=true;
}

if(		(
			(	!empty($_GET['getf'])							)||
			(	!empty($_GET['editf']) && isset($_POST['fdata']))||
			(	!empty($_GET['delf'])							)||
			(	!empty($_GET['delfld'])							)||
			(	!empty($_GET['chmod'])							)||
			(	isset($_FILES['file']) && isset($_POST['dir'])	)||
			(	!empty($_GET['copy'])							)||
			(	!empty($_GET['mkdir'])							)
		) && !(
			$auth['loggedin']
		)
){
	echo 'Sorry, but you can\'t modify files if you\'re not logged in!';
	die();
}

$reqs=@array($_GET['editf'],$_GET['delf'],$_GET['delfld'],$_GET['to'],$_POST['dir'].$_FILES['file']['name'],$_GET['download']);
foreach($reqs as $req){
	if( !empty($req) ){
		$reqq=str_replace('\\','/',realpath($req));
		foreach($inv as $in){
			if(substr($reqq,0,strlen($in)) == $in){
				$ret=array('status'=>100);
				echo $_GET['cb']."(".json_encode($ret).");";
				die();
			}
		}
	}
}

//							 _       _          __ _ _
//			 _   _ _ __   __| | __ _| |_ ___   / _(_) | ___  ___
//			| | | | '_ \ / _` |/ _` | __/ _ \ | |_| | |/ _ \/ __|
//			| |_| | |_) | (_| | (_| | ||  __/ |  _| | |  __/\__ \
//			 \__,_| .__/ \__,_|\__,_|\__\___| |_| |_|_|\___||___/
//				  |_|
//

if(isset($_GET['d'])){
	$files=array();
	$dirs=array();
	$dirch=true;
	if(!empty($_GET['d'])){
		$cdir=stripslashes(rawurldecode($_GET['d']));
		//$cdir=unslash(str_replace('\\','/',$cdir));
		while( ($cdir[strlen($cdir) -1 ]) == '/' ){
			$cdir=substr($cdir,0,-1);
		}
		$dirch=@chdir($cdir);
	}else{
		$dirch=@chdir('../');
		$cdir=str_replace('\\','/',getcwd());
	}
	$cdirs=explode('/',$cdir);
	if($cdirs[0]==''){
		$cdirs[0]='/';
	}
	if( $dirch && ($auth['loggedin'] || substr(unslash($cdir),0,strlen($path))==$path) && (realpath($cdir)!=dirname($_SERVER['SCRIPT_FILENAME'])) ){ 
		if(empty($cdirs[count($cdirs)-1])) array_pop($cdirs);
		$tfiles=glob('*');
		if(is_array($tfiles) && count($tfiles) > 0){
			foreach($tfiles as $tfile){
				if(is_dir($tfile)){
					if(!in_array($cdir.'/'.$tfile,$inv)){
						$stat=@stat($tfile);
						$ui=posix_getpwuid($stat['uid']);
						$gi=posix_getgrgid($stat['gid']);
						$dirs[]=array('n'=>stripslashes($tfile), 's'=>'', 'fp'=>view_perms_color($tfile), 'p'=>view_perms(@fileperms($tfile)));
					}
				}else{
					if(!in_array($cdir.'/'.$tfile,$inv)){
						$fext=substr($tfile,strrpos($tfile,'.')+1);
						$type='page';
						foreach($ext as $name=>$exts){
							if(in_array(strtolower($fext),$exts)) $type=$name;
						}
						$stat=@stat($tfile);
						$ui=posix_getpwuid($stat['uid']);
						$gi=posix_getgrgid($stat['gid']);
						$bin=is_binary($tfile) ? 'true' : 'false';
						$files[]=array('n'=>stripslashes($tfile), 's'=>size_readable(@filesize($tfile)), 'fp'=>view_perms_color($tfile), 'p'=>view_perms(@fileperms($tfile)), 't'=>$type, 'b'=>$bin);
					}
				}
			}
		}
	}
	$arr=array('jsonlock'=>$_GET['lock'],'cdir'=>$cdir,'cdirs'=>$cdirs,'dirs'=>$dirs,'files'=>$files);
	echo $_GET['cb']."(".json_encode($arr).");";
	die();
}

//							_              _          __  __
//			 _ __ ___   ___| |_ __ _   ___| |_ _   _ / _|/ _|
//			| '_ ` _ \ / _ \ __/ _` | / __| __| | | | |_| |_
//			| | | | | |  __/ || (_| | \__ \ |_| |_| |  _|  _|
//			|_| |_| |_|\___|\__\__,_| |___/\__|\__,_|_| |_|
//

if(!empty($_GET['getinfo'])){
	// status codes:
	// 0 	ok
	// 1	file doesn't exist
	$f=stripslashes(rawurldecode($_GET['getinfo']));
	$ret=array('status'=>0);
	if(!file_exists($f)){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	$stat=@stat($f);
	$ui=posix_getpwuid($stat['uid']);
	$gi=posix_getgrgid($stat['gid']);
	$ret['f']=$stat;
	$ret['u']=$ui;
	$ret['g']=$gi;
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

if(isset($_GET['lt'])){
	if($win){
		$lt=array();
		foreach(range('c','z') as $ltr){
			if(is_dir($ltr.':/')) $lt[]=strtoupper($ltr);
		}
		echo $_GET['cb']."(".json_encode($lt).");";
	}
	die();
}

//					  _ _ _      __ _ _
//			  ___  __| (_) |_   / _(_) | ___
//			 / _ \/ _` | | __| | |_| | |/ _ \
//			|  __/ (_| | | |_  |  _| | |  __/
//			 \___|\__,_|_|\__| |_| |_|_|\___|
//	

if(!empty($_GET['getf'])){
	// status codes:
	// 0 	ok
	// 1	file doesn't exist
	// 2	file is binary
	$f=stripslashes(rawurldecode($_GET['getf']));
	$ret=array('status'=>0,'data'=>'');
	if(!file_exists($f)){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(is_binary($f)){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(filesize($f) > 0){
		$fh = fopen($f, "rb");
		$file = fread($fh,filesize($f));
		fclose($fh);
		clearstatcache();
		$ret['data']=$file;
	}
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

if(!empty($_GET['editf']) && isset($_POST['fdata']) ){
	// status codes:
	// 0 	ok
	// 1	file isn't writable
	// 2	invalid filename
	// 3	couldn't write to file
	$f=stripslashes(rawurldecode($_GET['editf']));
	if(get_magic_quotes_gpc()){
		$data=stripslashes($_POST['fdata']);
	}else{
		$data=$_POST['fdata'];
	}
	$ret=array('status'=>0);
	if(file_exists($f)){
		if(!is_writable($f)){
			$ret['status']=1;
			echo $_GET['cb']."(".json_encode($ret).");";
			die();
		}
	}else{
		if(!is_writable(dirname($f))){
			$ret['status']=1;
			echo $_GET['cb']."(".json_encode($ret).");";
			die();
		}
	}
	$invalid=false;
	foreach(str_split(basename($f)) as $ff){
		if(in_array($ff,$fname_blacklist)){
			$invalid=true;
		}
	}
	if($invalid){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	$fh = @fopen($f, "wb");
	if(!$fh){
		$ret['status']=3;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	fwrite($fh,$data);
	fclose($fh);
	clearstatcache();
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

//				 _      _      _          __ _ _
//			  __| | ___| | ___| |_ ___   / _(_) | ___
//			 / _` |/ _ \ |/ _ \ __/ _ \ | |_| | |/ _ \
//			| (_| |  __/ |  __/ ||  __/ |  _| | |  __/
//			 \__,_|\___|_|\___|\__\___| |_| |_|_|\___|
//

if(!empty($_GET['delf'])){
	// status codes:
	// 0 	ok
	// 1	file doesn't exist
	// 2	file isn't writable
	// 3	couldn't delete
	$f=stripslashes(rawurldecode($_GET['delf']));
	$ret=array('status'=>0);
	if(!file_exists($f)){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!is_writable($f)){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!@unlink($f)){
		$ret['status']=3;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

//					  _ _ _      __       _     _
//			  ___  __| (_) |_   / _| ___ | | __| | ___ _ __
//			 / _ \/ _` | | __| | |_ / _ \| |/ _` |/ _ \ '__|
//			|  __/ (_| | | |_  |  _| (_) | | (_| |  __/ |
//			 \___|\__,_|_|\__| |_|  \___/|_|\__,_|\___|_|
//

if(!empty($_GET['mkdir'])){
	// status codes:
	// 0 	ok
	// 1	parent isn't writable
	// 2	invalid filename
	// 3	couldn't write to file
	$f=stripslashes(rawurldecode($_GET['mkdir']));
	$ret=array('status'=>0);
	if(!is_writable(dirname($f))){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	$invalid=false;
	foreach(str_split(basename($f)) as $ff){
		if(in_array($ff,$fname_blacklist)){
			$invalid=true;
		}
	}
	if($invalid){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!@mkdir($f)){
		$ret['status']=3;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

//				 _      _      _          __       _     _
//			  __| | ___| | ___| |_ ___   / _| ___ | | __| | ___ _ __
//			 / _` |/ _ \ |/ _ \ __/ _ \ | |_ / _ \| |/ _` |/ _ \ '__|
//			| (_| |  __/ |  __/ ||  __/ |  _| (_) | | (_| |  __/ |
//			 \__,_|\___|_|\___|\__\___| |_|  \___/|_|\__,_|\___|_|
//

if(!empty($_GET['delfld'])){
	// status codes:
	// 0 	ok
	// 1	file doesn't exist
	// 2	file isn't writable
	// 3	couldn't delete
	$f=stripslashes(rawurldecode($_GET['delfld']));
	$ret=array('status'=>0);
	if(!file_exists($f)){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!is_writable($f)){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!@deleteAll($f)){
		$ret['status']=3;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

//						 _                 _    __ _ _
//			 _   _ _ __ | | ___   __ _  __| |  / _(_) | ___
//			| | | | '_ \| |/ _ \ / _` |/ _` | | |_| | |/ _ \
//			| |_| | |_) | | (_) | (_| | (_| | |  _| | |  __/
//			 \__,_| .__/|_|\___/ \__,_|\__,_| |_| |_|_|\___|
//				  |_|
//

if( isset($_FILES['file']) && isset($_POST['dir']) ){
	// status codes:
	// 0 	ok
	// 1	file already exists
	// 2	couldn't upload
	
	$dest=$_POST['dir'].'/';
	$where=$dest.basename($_FILES['file']['name']);
	
	$ret=array('status'=>0);
	if( file_exists($where) && !isset($_POST['force']) ){
		$ret['status']=1;
		$ret['type']=is_dir($where) ? 'dir' : 'file';
	}elseif(!@move_uploaded_file($_FILES['file']['tmp_name'], $where)){
		$ret['status']=2;
	}
	
	?>
	<script>
		window.top.window.stopUpload(<?php echo json_encode($ret); ?>);
	</script>
	<?php
	die();
}

//			  __ _ _             _               _
//			 / _(_) | ___    ___| |__   ___  ___| | __
//			| |_| | |/ _ \  / __| '_ \ / _ \/ __| |/ /
//			|  _| | |  __/ | (__| | | |  __/ (__|   <
//			|_| |_|_|\___|  \___|_| |_|\___|\___|_|\_\
//

if(!empty($_GET['fcheck'])){
	$ret=array();
	$f=stripslashes(rawurldecode($_GET['fcheck']));
	$ret['exists']=@file_exists($f);
	$ret['type']=@is_dir($f) ? 'dir' : 'file';
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

//									  __ _ _
//			  ___ ___  _ __  _   _   / _(_) | ___
//			 / __/ _ \| '_ \| | | | | |_| | |/ _ \
//			| (_| (_) | |_) | |_| | |  _| | |  __/
//			 \___\___/| .__/ \__, | |_| |_|_|\___|
//					  |_|    |___/
//

if(!empty($_GET['copy']) && $_GET['copy']=='copy'){
	// status codes:
	// 0 	ok
	// 1	no file
	// 2	dest not writable
	// 3	invalid fname
	// 4	couldn't
	$f=stripslashes(rawurldecode($_GET['from']));
	$to=stripslashes(rawurldecode($_GET['to']));
	$ret=array('status'=>0);
	if(!file_exists($f)){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!is_writable(dirname($to))){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	$invalid=false;
	foreach(str_split(basename($to)) as $ff){
		if(in_array($ff,$fname_blacklist)){
			$invalid=true;
		}
	}
	if($invalid){
		$ret['status']=3;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!@copy($f,$to)){
		$ret['status']=4;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

if(!empty($_GET['copy']) && $_GET['copy']=='move'){
	// status codes:
	// 0 	ok
	// 1	no file
	// 2	origin not writable
	// 3	dest not writable
	// 4	invalid filename
	// 5	couldn't
	$f=stripslashes(rawurldecode($_GET['from']));
	$to=stripslashes(rawurldecode($_GET['to']));
	$ret=array('status'=>0);
	if(!file_exists($f)){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!is_writable($f)){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!is_writable(dirname($to))){
		$ret['status']=3;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	$invalid=false;
	foreach(str_split(basename($to)) as $ff){
		if(in_array($ff,$fname_blacklist)){
			$invalid=true;
		}
	}
	if($invalid){
		$ret['status']=4;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!@rename($f,$to)){
		$ret['status']=5;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

//				 _                     _                 _    __ _ _
//			  __| | _____      ___ __ | | ___   __ _  __| |  / _(_) | ___
//			 / _` |/ _ \ \ /\ / / '_ \| |/ _ \ / _` |/ _` | | |_| | |/ _ \
//			| (_| | (_) \ V  V /| | | | | (_) | (_| | (_| | |  _| | |  __/
//			 \__,_|\___/ \_/\_/ |_| |_|_|\___/ \__,_|\__,_| |_| |_|_|\___|
//

if(!empty($_GET['download'])){
	$file=stripslashes(rawurldecode($_GET['download']));
	$name=basename($file);
	if( !$auth['loggedin'] && substr($file,0,strlen($path))!=$path ){ 
		echo 'Sorry, but you need to login to be able to download that file.';
		die();
	}
	if(!file_exists($file)){
		echo 'Sorry, the file you\'re trying to download doesn\'t exist.';
		die();
	}
	if(!is_readable($file)){
		echo 'Sorry, the file you\'re trying to download isn\'t readable.';
		die();
	}
	if(strpos($name,'.')!==FALSE){
		$fext=strtolower(substr($file,strrpos($file,'.')+1));
	}
	if(!$auth['loggedin'] && in_array($fext,$ext['php'])){
		echo 'Sorry, but you need to login to be able to download that file.';
		die();
	}
	$mime='application/octet-stream';
	if(!isset($_GET['forcedl'])){
		if(!empty($fext)){			
			if(		in_array($fext,$ext['php']) || in_array($fext,$ext['c']) || in_array($fext,$ext['cpp']) ||
					in_array($fext,$ext['cs']) || in_array($fext,$ext['css']) || in_array($fext,$ext['html']) ||
					in_array($fext,$ext['xml']) || in_array($fext,$ext['js']) || in_array($fext,$ext['sql']) ||
					in_array($fext,$ext['rb']) || in_array($fext,$ext['code']) || in_array($fext,$ext['text'])
			){
				$mime='text/plain';
			}else if(in_array($fext,$ext['img'])){
				$mime='image/'.str_replace('jpg','jpeg',str_replace('apng','png',$fext));
			}else if(in_array($fext,$ext['audio'])){
				switch($fext){
					case 'mp3':
						$mime='audio/mpeg';
						break;
					case 'flac':
						$mime='audio/flac';
						break;
					case 'aac':
						$mime='audio/aac';
						break;
					case 'm4a':
						$mime='audio/mp4a-latm';
						break;
					case 'mpga':
						$mime='audio/mpeg';
						break;
					case 'ogg':
						$mime='audio/ogg';
						break;
				}
			}
		}else if(!is_binary($file)){
			$mime='text/plain';
		}
	}
	if($mime=='application/octet-stream'){
		header('Content-Disposition: attachment; filename="'.$name.'"');
	}else{
		header('Content-Disposition: inline; filename="'.$name.'"');
	}
	header('Content-Description: File Transfer');
	header('Content-Type: '.$mime);
	header('Content-Transfer-Encoding: binary');
	header('Expires: 0');
	header('Cache-Control: must-revalidate, post-check=0, pre-check=0');
	header('Pragma: public');
	header('Content-Length: ' . filesize($file));
	ob_clean();
	flush();
	readfile($file);
	die();
}

//						_         _
//			  __ _  ___| |_   ___(_)_______
//			 / _` |/ _ \ __| / __| |_  / _ \
//			| (_| |  __/ |_  \__ \ |/ /  __/
//			 \__, |\___|\__| |___/_/___\___|
//			 |___/
//

if(!empty($_GET['getsize'])){
	// status codes:
	// 0 	ok
	// 1	no file
	// 2	couldn't get size, probably not an image
	$img=stripslashes(rawurldecode($_GET['getsize']));
	$ret=array('status'=>0);
	if(!file_exists($img)){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!list($width, $height, $i_type, $i_attr) = @getimagesize($img)){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	$ret['width']=$width;
	$ret['height']=$height;
	// get next and previous imgs
	$imgs=array();
	$files=glob(dirname($img).'/*');
	foreach($files as $key=>$file){
		if($file==$img) break; 
	}
	$found=false;
	$ret['imgb']='';
	for($i=$key-1;$i>=0 && !$found;$i--){
		$fname=$files[$i];
		if(strpos($fname,'.')!==FALSE){
			$fext=strtolower(substr($fname,strrpos($fname,'.')+1));
			if(in_array($fext,$ext['img'])){
				$ret['imgb']=$fname;
				$found=true;
			}
		}
	}
	$found=false;
	$ret['imga']='';
	for($i=$key+1;$i<=count($files)-1 && !$found;$i++){
		$fname=$files[$i];
		if(strpos($fname,'.')!==FALSE){
			$fext=strtolower(substr($fname,strrpos($fname,'.')+1));		
			if(in_array($fext,$ext['img'])){
				$ret['imga']=$fname;
				$found=true;
			}
		}
	}
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

//				  _                         _
//			  ___| |__  _ __ ___   ___   __| |
//			 / __| '_ \| '_ ` _ \ / _ \ / _` |
//			| (__| | | | | | | | | (_) | (_| |
//			 \___|_| |_|_| |_| |_|\___/ \__,_|
//

if(!empty($_GET['chmod']) && !empty($_GET['to'])){
	// status codes:
	// 0 	ok
	// 1	file doesn't exist
	// 2	couldn't chmod
	$f=stripslashes(rawurldecode($_GET['chmod']));
	$to=$_GET['to'];
	$ret=array('status'=>0);
	if(!file_exists($f)){
		$ret['status']=1;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	if(!@chmod($f,octdec('0'.$to))){
		$ret['status']=2;
		echo $_GET['cb']."(".json_encode($ret).");";
		die();
	}
	echo $_GET['cb']."(".json_encode($ret).");";
	die();
}

echo 'Unknown action.';























index.php
<?php

// Tastydir is copyright Vlad Harbuz (vladh.net)

define('TASTYDIR_VERSION','1.2');
define('TASTYDIR_SUBVERSION','1251');

error_reporting(E_ALL);

if(!file_exists('_tastydir/settings.php')){
	require_once('_tastydir/install.php');
	die();
}

require_once '_tastydir/common.php';
require_once '_tastydir/settings.php';

//////////////////////////
//	authentication
////////
$auth['loggedin']=false;

if($auth['use']){

	// login action
	if(isset($_POST['action']) && $_POST['action']=='login' && !empty($_POST['pw'])){
		if(sha1($_POST['pw'])==$auth['pass']){
			// set cookie for 24 hours
			setcookie('tastydir_auth',sha1($auth['pass']),time()+86400);
			header('Location: index.php');
		}
	}

	// logout action
	if(isset($_GET['action']) && $_GET['action']=='logout'){
		setcookie('tastydir_auth','',time()-86400);
		header('Location: index.php');
	}

	// check if logged in
	if(!empty($_COOKIE['tastydir_auth']) && $_COOKIE['tastydir_auth']==sha1($auth['pass'])){
		$auth['loggedin']=true;
	}
}else{
	$auth['loggedin']=true;
}

?>
<!doctype html>
<html>
	<head>
		<title>tastydir</title>
		<link rel="stylesheet" href="_tastydir/style.css">
	</head>
	<body>
		<div id="everything" class="fluid">
			<div id="header">
				<h1>tastydir</h1>
			</div>
			<div id="nav" class="rounded_top clearfix">
				<div id="navnav">
					<button class="icon" onClick="history.go(-1);">&laquo;</button>
					<button class="icon" onClick="updateFiles(document.location.hash.substr(1));"><img src="_tastydir/images/arrow_refresh.png" alt="Refresh" title="Refresh"></button>
					<button onclick="document.location='#';">Home</button>
					<?php if($win){ ?>
					<div id="letters">
						<button class="icon">
							<img src="_tastydir/images/loading.gif">
						</button>
					</div>
					<?php } ?>
				</div>
				<?php if($auth['use']){ ?>
				<div id="meta">
					<div class="item"><a rel="tipsy" id="topinfo"><img src="_tastydir/images/information.png" alt="Info"></a></div>
					<div class="sep"></div>
					<?php
					if($auth['loggedin']){ ?>
						<div class="item" id="login"><a href="index.php?action=logout">Log out <img src="_tastydir/images/door_out.png" title="Log out" alt="Log out"></a></div>
					<?php }else{ ?>
						<div class="item" id="login"><a href="javascript:;" onClick="replaceLogin();">Log in <img src="_tastydir/images/user_go.png" title="Log in" alt="Log in"></a></div>
					<?php } ?>
				</div>
				<?php } ?>
			</div>
			<div id="path" class="clearfix">
				
			</div>
			<div id="body">
				<div id="files_outer">
					<table id="files_table">
						<thead id="files_head">
							<tr>
								<th style="width:55%;">Name</th>
								<th style="width:15%;">Size</th>
								<th style="width:15%;">Permissions</th>
								<th style="width:15%;">Actions</th>
							</tr>
						</thead>
						<tbody id="files">
						</tbody>
					</table>
				</div>
				<div class="modal" id="loading_modal">
					<img src="_tastydir/images/loading.gif">
				</div>
				<div class="modal" id="editfile_modal">
				</div>
			</div>
			<div id="bottombar" class="clearfix">
				<div id="bottombar_l">Navigate to the folder you want to copy or move <span id="copyname"></span> to, then use the buttons on the right.</div>
				<div id="bottombar_r">
					<a href="javascript:;" onClick="closeCopyBar();">Cancel <img src="_tastydir/images/cancel.png" alt="Cancel" title="Cancel"></a>
					<span style="padding:0 10px;">&nbsp;</span> <!-- ok so this is hacky -->
					<a href="javascript:;" onClick="copyFileAttempt(cdir,false);">Copy here <img src="_tastydir/images/folder_go.png" alt="Cancel" title="Cancel"></a>
					<span style="padding:0 10px;">&nbsp;</span> <!-- ok so this is hacky -->
					<a href="javascript:;" onClick="copyFileAttempt(cdir,true);">Move here <img src="_tastydir/images/folder_go.png" alt="Cancel" title="Cancel"></a>
				</div>
			</div>
			<div id="footer" class="rounded_bottom clearfix">
				<div id="makefiles">
					<div class="makefile"><a href="javascript:;" onClick="createFileDialog(cdir);">Create file <img src="_tastydir/images/page_white_add.png" alt="Create file" title="Create file"></a></div>
					<div class="sep"></div>
					<div class="makefile"><a href="javascript:;" onClick="createFolderDialog(cdir);">Create folder <img src="_tastydir/images/folder_add.png" alt="Create folder" title="Create folder"></a></div>
				</div>
				<div id="uploadfile">
					<div id="uploadform">
						<div id="uploadform_progress"><img src="_tastydir/images/loading.gif"></div>
						<div id="uploadform_really">
							<form enctype="multipart/form-data" action="_tastydir/do.php" method="POST" onsubmit="return startUpload();" target="hacky_iframe" id="uploadform_seriously">
								<input name="file" type="file" id="upload_finput">
								<input type="submit" value="Upload">
							</form>
						</div>
						<iframe id="hacky_iframe" name="hacky_iframe" src="#"></iframe> 
					</div>
					<div id="uploadtext">
						<a href="javascript:;" onClick="showUploadForm();"><span id="uploadtext_really">Upload file</span> <img src="_tastydir/images/page_white_get.png" alt="Upload file" title="Upload file"></a>
					</div>
				</div>
			</div>
		</div>
		<script>
			var t_dn='<?php echo $t_dn; ?>';
			var loggedin=<?php echo $auth['loggedin'] ? 'true' : 'false'; ?>;
			var t_version='<?php echo TASTYDIR_VERSION; ?>';
			var t_subversion=<?php echo TASTYDIR_SUBVERSION; ?>;
			var t_fname_blacklist='<?php foreach($fname_blacklist as $key=>$ff){ if($key!=(count($fname_blacklist)-1)){ if($key!=0){echo ' ';} echo str_replace('\\','\\\\',$ff); } } ?>';
			var t_metainfo = {
				user: '<?php echo $t_curruser; ?>'
			};
			var t_switchlayout=false;
			var t_path='<?php echo $path; ?>';
		</script>
		<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.4.2/jquery.min.js"></script>
		<script src="http://ajax.googleapis.com/ajax/libs/jqueryui/1.8.2/jquery-ui.min.js"></script>
		<script src="_tastydir/js/jquery.ba-hashchange.min.js"></script>
		<script src="_tastydir/js/jquery.tipsy.js"></script>
		<script src="_tastydir/js/site.js"></script>
	</body>
</html>






















